From 10a17e71bd79686b8ba5fa5d6f33400cc886959b Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Tue, 09 Nov 2021 20:30:24 +0200
Subject: [PATCH] [SW-56233] gaudi2: security: prot bits init update

Currently in preboot as per the LBW spec update we need
to support more than 1 protection bit configuration index.
This patch extends the existing infra to support the same.

Change-Id: Ibe09a6f55059ff8cd3e8ab2c253034e0226735d8
---

diff --git a/arch/gaudi2/common/security.c b/arch/gaudi2/common/security.c
index 6dfa14f..5dd97c9 100644
--- a/arch/gaudi2/common/security.c
+++ b/arch/gaudi2/common/security.c
@@ -383,9 +383,9 @@
 		 mmPSOC_ARC1_MSTR_IF_CORE_LBW_AXPROT_OVRD);
 }
 
-static int init_prot_bits_block(u8 major, u8 minor, u8 sub_minor)
+static int init_prot_bits_block(u32 index, u8 major, u8 minor, u8 sub_minor)
 {
-	struct hl_prot_bits_cfg *cfg = &gaudi2_prot_bits_cfg[0];
+	struct hl_prot_bits_cfg *cfg = &gaudi2_prot_bits_cfg[index];
 	u8 b_idx = 0, d_idx = 0;
 	u32 val;
 	u32 addr, reg_addr;
@@ -419,19 +419,23 @@
 
 static void hl_dev_prot_bits_init(void)
 {
-	u8 major_idx, minor_idx, sub_minor_idx;
+	u8 major_idx, minor_idx, sub_minor_idx, cfg_idx;
 	struct hl_prot_bits_addr block_addr;
+	u32 conf_sz = ARRAY_SIZE(gaudi2_prot_bits_cfg);
 
-	block_addr = gaudi2_prot_bits_cfg[0].addr;
-	for (major_idx = 0; major_idx < block_addr.major; major_idx++) {
-		minor_idx = 0;
-		do {
-			sub_minor_idx = 0;
+	for (cfg_idx = 0; cfg_idx < conf_sz; cfg_idx++) {
+		block_addr = gaudi2_prot_bits_cfg[cfg_idx].addr;
+		for (major_idx = 0; major_idx < block_addr.major; major_idx++) {
+			minor_idx = 0;
 			do {
-				init_prot_bits_block(major_idx, minor_idx, sub_minor_idx);
-			} while (sub_minor_idx < block_addr.sub_minor);
-			minor_idx++;
-		} while (minor_idx < block_addr.minor);
+				sub_minor_idx = 0;
+				do {
+					init_prot_bits_block(cfg_idx, major_idx, minor_idx,
+							     sub_minor_idx);
+				} while (sub_minor_idx < block_addr.sub_minor);
+				minor_idx++;
+			} while (minor_idx < block_addr.minor);
+		}
 	}
 
 	preboot_puts("gaudi2 protection bits configured\n");
